ipld/src/util.rs

```diff
@@ -75,46 +87,82 @@ where
     let bytes = load_block(root).await?;
     let ipld = from_slice(&bytes)?;
 
-    traverse_ipld_links_hash(walked, load_block, &ipld).await?;
+    traverse_ipld_links_hash(walked, load_block, &ipld, on_inserted).await?;
 
     Ok(())
 }
 
+lazy_static! {
+    pub static ref WALK_SNAPSHOT_PROGRESS_EXPORT: Arc<(AtomicU64, AtomicU64)> =
+        Arc::new((AtomicU64::new(0), AtomicU64::new(0)));
+    pub static ref WALK_SNAPSHOT_PROGRESS_DB_GC: Arc<(AtomicU64, AtomicU64)> =
+        Arc::new((AtomicU64::new(0), AtomicU64::new(0)));
+}
+
 /// Walks over tipset and state data and loads all blocks not yet seen.
 /// This is tracked based on the callback function loading blocks.
 pub async fn walk_snapshot<F, T>(
     tipset: &Tipset,
     recent_roots: i64,
     mut load_block: F,
+    progress_bar_message: Option<&str>,
+    progress_tracker: Option<Arc<(AtomicU64, AtomicU64)>>,
+    estimated_total_records: Option<u64>,
 ) -> anyhow::Result<usize>
 where
     F: FnMut(Cid) -> T + Send,
     T: Future<Output = anyhow::Result<Vec<u8>>> + Send,
 {
+    let estimated_total_records = estimated_total_records.unwrap_or_default();
+    let bar = ProgressBar::new(estimated_total_records);
+    bar.message(progress_bar_message.unwrap_or("Walking snapshot "));
+    bar.set_units(progress_bar::Units::Default);
+    bar.set_max_refresh_rate(Some(Duration::from_millis(500)));
+
     let mut seen = CidHashSet::default();
     let mut blocks_to_walk: VecDeque<Cid> = tipset.cids().to_vec().into();
     let mut current_min_height = tipset.epoch();
     let incl_roots_epoch = tipset.epoch() - recent_roots;
 
+    let on_inserted = {
+        let bar = bar.clone();
+        let progress_tracker = progress_tracker.clone();
+        move |len: usize| {
+            let progress = len as _;
+            let total = if progress > estimated_total_records {
+                progress
+            } else {
+                estimated_total_records
+            };
```

Comment:
I believe this trick should work
```suggestion
            let total = progress.max(estimated_total_records);
```